core/silkworm/consensus/clique/clique_snapshot.cpp

```diff
@@ -0,0 +1,248 @@
+    }
+
+    // We check what the vote is
+    bool authorize;
+    if (header.nonce == kNonceAuthorize) {
+        authorize = true;
+    } else if (header.nonce == kNonceUnauthorize) {
+        authorize = false;
+    } else {
+        return ValidationResult::kInvalidVote;
+    }
+    
+    tmp_recents.push_front(signer);
+    recents_ = tmp_recents;
+
+    if (header.beneficiary == 0x0000000000000000000000000000000000000000_address) {        
+        update(header.number, hash);
+        return ValidationResult::kOk;
+    }
+
+    // Uncast votes from signers
+    uncast(header.beneficiary, signer);
+    // do casting
+    cast(header.beneficiary, signer, authorize);
+    // If the vote passed, update the list of signers
+    auto current_tally{tallies_[header.beneficiary]};
+    if (current_tally.votes > signers_.size() / 2) {
+        if (current_tally.authorize) {
+            signers_.push_back(header.beneficiary);
+        } else {
+            signers_.erase(std::remove(signers_.begin(), signers_.end(), header.beneficiary), signers_.end());
+            // Clean up recents
+            recents_.pop_back();
+            // Update Tallies
+            uncast_all(header.beneficiary);
+        }
+        if (tallies_.count(header.beneficiary)) {
+            tallies_.erase(header.beneficiary);
+        }
+    }
+
+    update(header.number, hash);
+    // Success
+    return ValidationResult::kOk;
+}
+
+//! \brief Verify seal for header
+//! \param header: header to verify.
+ValidationResult CliqueSnapshot::verify_seal(BlockHeader header, const evmc::address& signer) {
+    // Check difficituly
+    auto authority{is_authority(header.number, signer)};
+    if (authority && header.difficulty != kDiffInTurn && header.difficulty != kDiffNoTurn) {
+        return ValidationResult::kIntrinsicGas;
+    }
+
+    return ValidationResult::kOk;
+}
+
+//! \brief Checks for authority
+//! \param block_number: Block to check.
+//! \param address: Address to check.
+//! \return if a signer at a given block height is in charge or not.
+bool CliqueSnapshot::is_authority(uint64_t block_number, evmc::address address) const noexcept {
+    uint64_t offset{0};
+	while (offset < signers_.size() && signers_[offset] != address) {
+		++offset;
+	}
+	return (block_number % signers_.size()) == offset;
+}
+
+//! \brief Getter method for signers_.
+//! \return Snapshot's signers.
+const std::vector<evmc::address>& CliqueSnapshot::get_signers() const noexcept {
+    return signers_;
+}
+
+//! \brief Getter method for block_number_.
+//! \return Snapshot's block number.
+const uint64_t& CliqueSnapshot::get_block_number() const noexcept {
+    return block_number_;
+}
+
+//! \brief Getter method for hash_.
+//! \return Snapshot's hash.
+const evmc::bytes32& CliqueSnapshot::get_hash() const noexcept {
+    return hash_;
+}
+
+//! \brief Convert the snapshot in JSON.
+//! \return The resulting JSON.
+Bytes CliqueSnapshot::to_bytes() const noexcept {
```

Comment:
Seems to me the comment is misleading. Says it returns JSON bu the function returns Bytes